fedimint-api/src/db/mod.rs

```diff
@@ -134,21 +138,90 @@ pub trait IDatabaseTransaction<'a>: 'a + Send {
 
 // TODO: use macro again
 #[doc = " A handle to a type-erased database implementation"]
-pub struct DatabaseTransaction<'a>(
+pub struct ReadOnlyDatabaseTransaction<'a>(
     Box<dyn IDatabaseTransaction<'a> + Send + 'a>,
     ModuleDecoderRegistry,
 );
+
+impl<'a> ReadOnlyDatabaseTransaction<'a> {
+    pub fn new<I: IDatabaseTransaction<'a> + Send + 'a>(
+        dbtx: I,
+        decoders: ModuleDecoderRegistry,
+    ) -> ReadOnlyDatabaseTransaction<'a> {
+        ReadOnlyDatabaseTransaction(Box::new(dbtx), decoders)
+    }
+
+    pub async fn get_value<K>(&mut self, key: &K) -> Result<Option<K::Value>>
+    where
+        K: DatabaseKey + DatabaseKeyPrefixConst,
+    {
+        let key_bytes = key.to_bytes();
+        let value_bytes = match self.0.raw_get_bytes(&key_bytes).await? {
+            Some(value) => value,
+            None => return Ok(None),
+        };
+
+        trace!(
+            "get_value: Decoding {} from bytes {:?}",
+            std::any::type_name::<K::Value>(),
+            value_bytes
+        );
+        Ok(Some(K::Value::from_bytes(&value_bytes, &self.1)?))
+    }
+
+    pub async fn find_by_prefix<KP>(
+        &mut self,
+        key_prefix: &KP,
+    ) -> impl Iterator<Item = Result<(KP::Key, KP::Value)>> + '_
+    where
+        KP: DatabaseKeyPrefix + DatabaseKeyPrefixConst,
+    {
+        let decoders = self.1.clone();
+        let prefix_bytes = key_prefix.to_bytes();
+        self.0
+            .raw_find_by_prefix(&prefix_bytes)
+            .await
+            .map(move |res| {
+                res.and_then(|(key_bytes, value_bytes)| {
+                    let key = KP::Key::from_bytes(&key_bytes, &decoders)?;
+                    trace!(
+                        "find by prefix: Decoding {} from bytes {:?}",
+                        std::any::type_name::<KP::Value>(),
+                        value_bytes
+                    );
+                    let value = KP::Value::from_bytes(&value_bytes, &decoders)?;
+                    Ok((key, value))
+                })
+            })
+    }
+}
+
+pub struct CommitTracker {
+    is_committed: bool,
+}
+
+impl Drop for CommitTracker {
+    fn drop(&mut self) {
+        if !self.is_committed {
+            warn!("DatabaseTransaction has not called commit.");
+        }
+    }
+}
+
+#[doc = " A handle to a type-erased database implementation"]
+pub struct DatabaseTransaction<'a>(ReadOnlyDatabaseTransaction<'a>, CommitTracker);
```

Comment:
I'm lost here. So "database transaction" wraps "read-only database transaction"? Seems weird. What's the relationship between things here and how is it supposed to work?

I kind of expected... not sure what exactly. :D

Maybe

```
pub struct RODatabaseTransaction<'a>{
    dyn_tx: Box<dyn IDatabaseTransaction<'a> + Send + 'a>,
    registry: ModuleDecoderRegistry,
    commit_tracker: CommitTracker,
}
```

where "being read-only" would only call something on `CommitTracker` to make it silent?

Do we aim at any type-level enforcement of actual ro vs rw? Like:

```
pub struct RODatabaseTransaction<'a>{
    dyn_tx: Box<dyn IDatabaseTransaction<'a> + Send + 'a>,
    registry: ModuleDecoderRegistry,
}

pub struct RWDatabaseTransaction<'a>{
    dyn_tx: Box<dyn IDatabaseTransaction<'a> + Send + 'a>,
    registry: ModuleDecoderRegistry,
    commit_tracker: CommitTracker,
}

trait RWDatabaseTransactionOps : RODatabaseTransaction { ... }

impl RODatabaseTransactionOps for RWDatabaseTransaction { ... }
impl RODatabaseTransactionOps for RWDatabaseTransaction { ... }
impl RWDatabaseTransactionOps for RWDatabaseTransaction { ... }

```

Seems kind of a lot infra, if all we needed was just warnings about uncommited txes.

Reply:
Haha yeah :) So currently DatabaseTransaction does wrap ReadOnlyDatabaseTransaction. I did this so I didn't need to re-write the logic for get_value and find_by_prefix. I had originally separated RW and R transactions into different traits, but that ended up being more work than it was worth. I didn't want to re-write get_value and find_by_prefix, but both RW and R transactions need to be able to read, so instead I decided to wrap it. Open to suggestions if there's a clean way.

I had also considered adding a new trait, but (and correct me if I'm wrong) the point of adding the DatabaseTransaction struct was to get rid of the Box<dyn ....> of IDatabaseTransaction, and I think I would have needed to add that back in (or add another erased type) if I were to add another trait. 

Right now there aren't any guard rails or type enforcement, you could potentially use ReadOnlyDatabaseTransaction to write and commit some data, but you'd need to reach inside to grab the boxed IDatabaseTransaction struct. Ultimately this change is mainly for readability and debugability, so I thought just having DatabaseTransaction use a ReadOnlyDatabaseTransaction so that the get_value and find_by_prefix code didn't need to be repeated. Let me know what you think.

Reply:
> Open to suggestions if there's a clean way.

If we introduce distinct types for RW vs RO, then there will be places in the code (some common functions) that will need to work with either (depending on what the caller has), and we will need some method to generalize over them. That's why I'm tempted to have only one type (`DatabaseTransaction { inner: Box<....>, ... other fields ... }`, and the "this is only RO" is going to be only an advisory. **Maybe to make it non-confusing, we should just drop the whole "read only" and name it "commit not required".**

Reply:
So:

```rust
DatabaseTransaction {
 inner: Box<dyn ...>,
 commit_tracker: Option<CommitTracker>,
}
```

and then user code would be calling `db.begin_transaction()` or `db.begin_transaction().commit_not_required()` and at any stage `.commit_not_required()` will just get rid of `commit_tracker` and prevent the transaction from complaining about commit.

By getting rid of RO transaction concept we are not conflating the two, and staying true to what we are after: just lettting devs know they've forgot about `commit_tx`.

If we ever decide that we want the full RO vs RW distinction (possibly to enforce certain behavior from modules), we can make RO imply `commit_not_required`.

Reply:
I like that idea. Also solves the problem of not re-implementing logic for RO transactions. Will make the change.

Reply:
I think I have a slightly cleaner solution that requires less changes. Since we know when inserts/deletes happen, we can just create the commit tracker then and default it to None. That way the warning message should only be emitted when there's a write to the transaction, otherwise the commit tracker will never exist.